daffodil-lib/src/main/scala/org/apache/daffodil/api/Validator.scala

```diff
@@ -0,0 +1,59 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.daffodil.api
+
+import com.typesafe.config.Config
+
+import scala.xml.SAXException
+
+trait Validator {
+  def validateXML(document: java.io.InputStream): ValidationResult
+}
+
+trait ValidatorFactory {
+  def name(): String
+  def make(config: Option[Config]): Validator
+}
+
+case class ValidationResult(warnings: Seq[ValidationWarning], errors: Seq[ValidationError]) {
+  def failed: Boolean = errors.nonEmpty
+}
+
+
+trait ValidationError {
+  def message: String
+  def toThrowable: Throwable
+}
+trait ValidationWarning
+// todo;; can validator return other metadata?
+//trait ValidationReport
+
+
+// todo;; can DRY this up after sanity check
+case class SaxValidationError(e: SAXException) extends ValidationError {
+  def message: String = e.getMessage
+  def toThrowable: Throwable = e
```

Comment:
The rationale here is that the `PState` needs a `Throwable` for `validationErrorNoContext`.  So this conversion is to (1) fit into that existing API and (2) not have `ValidationError` be throwable just for that reason.

Reply:
Typically what you have as "def toThrowable" would be "def cause". Java Exception objects have two members, a message and a cause. One of the two, or both, have to be defined. (Typically one of the two.) You're not extending java's Exception class here. I think probably you should be.

Reply:
> I think probably you should be.

Yeah for the sax results it makes sense to extend Exception, since they are Exception based anyways.  The point of the overall design though was to avoid forcing implementations of the new traits to be Exceptions.

Reply:
Reworked this some.  Added an extra trait for results that are exception based.  See what you think.